src/sources/raster.js

```diff
@@ -47,9 +47,14 @@ export class RasterTileSource extends NetworkTileSource {
 
     // Return texture info for a raster tile
     tileTexture (tile) {
-        let key = tile.coords.key;
+        let new_coords = tile.coords;
+        if (this.zoom_offset !== 0) {
```

Comment:
I think you can achieve this section (the calls to `coordinateAtZoom` and `coordinateWithMaxZoom`) with one call to `normalizedCoordinate`:

https://github.com/tangrams/tangram/blob/master/src/tile.js#L86

e.g.

`let coords = Tile.normalizedCoordinate(tile.coords, { max_zoom: this.max_zoom, zoom_bias: this.zoom_offset });`

(Side note, it looks like the third parameter for `normalizedCoordinate`, `style_zoom`, is vestigial/unused.)

Reply:
I'm still a bit confused thinking through the use of `zoom_bias` vs. `zoom_offset` here (sources can now have both). I think it's because we determined that attached raster sources ignore their `tile_size` and are required to use the `tile_size` of their parent vector source. This could be revisited. I think the true restriction is that attached rasters need a `tile_size` that is the **same or larger** than their parent, but **not smaller**. This is because of the way we sample rasters within vector parent source shaders, where the raster is mapped to a single GL texture sampler. The vector source can cover **all or part of** a single raster texture (in the case it's partially covered, UVs are automatically adjusted to cover the appropriate sub-rectangle for each vector tile), but cannot span multiple raster textures.

So:
- We could relax this so rasters allow an equal or larger `tile_size` to the parent vector (parent vector source throws an error when trying to attach a raster source with a smaller `tile_size`)
- Docs reflect this (I don't think we mention anything about the "locking" of `tile_size` for attached rasters right now.

These are notes for related revisions, but don't need to be part of this PR.

Reply:
I believe we're agreed here. As I understand it, `zoom_bias` is strictly internal, and tracks cases of overzoom past `max_zoom` limits. `zoom_offset` would be an additional external control, in cases where the user wants to load 1 or more zoom levels below the current map zoom level. `zoom_offset` would be applied if its result is a lower zoom than the result of applying`zoom_bias`.

Examples, assuming a map zoom of 11:

- `max_zoom` is 10 – `zoom_bias` is 1, requested zoom is 10.
- `max_zoom` is 10, `zoom_adjust` is 1 – `zoom_bias` and `zoom_adjust` are equal, requested zoom is 10.
- `max_zoom` is 11, `zoom_adjust` is 1 – `zoom_bias` of 0 is less than `zoom_adjust`, requested zoom is 10.
- `max_zoom` is 10, `zoom_adjust` is 2 – `zoom_bias` of 1 is less than `zoom_adjust`, requested zoom is 9.

Reply:
… this is slightly complicated by the fact that `zoom_bias` is also adjusted in cases where a datasource uses a non-256px tile, but I believe even then these manipulations should still work. https://github.com/tangrams/tangram/blob/e3cd42dad5a8e6416ceca69a10a1908bf46ad852/src/sources/data_source.js#L170

Reply:
I think there's still some confusion here.

`zoom_bias` arises from `tile_size` and the introduction of 512px tiles, and does something very similar to `zoom_offset` -- causes logically sized 256px tiles to load one or more levels "up". When we added 512px support, we opted to have an explicit pixel size param like `tile_size: 512` instead of making the user think in terms of zoom levels. But, `zoom_bias` isn't directly related to overzooming past `max_zoom`, though overzooming code does need to take it into account.

In fact, `zoom_offset` is incorporated into `zoom_bias` already (though in this current branch seems to have been mistakenly (?) referred to as `tile_adjust` [here](https://github.com/tangrams/tangram/pull/685/files#diff-06e3f25a9a00a1a8840d2720f9ab37e1R170)... and is also referred to as `zoom_adjust` in your comments above?). It's basically an extension of `zoom_bias`, and mostly doesn't need to be used again once the data source is created.

But we ran into a specific divergence from that here with attached rasters, because of their unique relationship to their parent vector source...

**For normal vector source and standalone raster source (not attached to a vector source) tile resolution**: `zoom_bias` incorporates `zoom_offset` already, and I think everything works as desired.

**For raster sources attached to a vector source**: `zoom_bias` of the raster source is (intentionally, but undocumented as far as I see) not being applied, because we have been forcing the `tile_size` (from which `zoom_bias` is derived) to inherit from the vector source (which is passing in the tile coords here, which will already have been adjusted for the vector source's `zoom_bias`). The additional code here then applies just the `zoom_offset` (instead of the whole `zoom_bias`), which otherwise would be skipped.

Does that all make sense? :)

My original comment was just that we can reuse existing code by using the existing `Tile.normalizedCoordinate` function, which works very similarly to the code in this block.

Reply:
Yep, thanks for catching that. I've addressed those misnomers.

Re: attached rasters: yes, I'm seeing this as well – in my experiments, the practical effect is that an attached raster effectively inherits the `zoom_offset` of its 'parent' vector datasource. Given the technical limitations you described, would it be possible  to allow an attached raster to load at a higher zoom level than its parent?

I'll take a look at that `normalizedCoordinate` function after I unwrap my head from around this wrinkle :)